### DISCUSSION TITLE
Individual control of motors of CF

### DISCUSSION THREAD

**Discussion by EJKT** [bitcraze/discussions#184] [Q&A] [answered]
Is it possible to give controls for the motor speeds themselves? I want to have only the back 2 motors spinning at a very low RPM so the CF slides forwards over the ground on a small contraption I made.

**Comment by knmcguire** [comment, marked as answer]
You could use the [motorPowerSet Parameter](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/params/#motorpowerset) group for that. First set a value for the back to motors (m2 and m3) and then put motorPowerSet.enable to True. 

Here are instructions to set parameters from the client: https://www.bitcraze.io/documentation/repository/crazyflie-clients-python/master/userguides/userguide_client/parameters_tab/

**Reply by EJKT** [reply]
Could I also do it with the python API? If so, I know how to set parameters but I'm not sure what the following restrictions for the motor power would translate to in python:
![image](https://user-images.githubusercontent.com/87431056/183619057-1ec6d18e-539d-4e51-ada9-6a2043a6e1a1.png)
In other words, what would the maximum parameter setting be?
Thanks

**Reply by knmcguire** [reply]
yes absolutely. Anything you can do with the cfclient you can do with the cflib. Check out the documentation for that.

Uint16_max stands for the maximum value of a unsigned 16 bit integer which is 65535. But the motors can already start turning lightly at a value of 3000

**Reply by NageshEranki** [reply]
Hi

I am interested in the same objective: to remotely control the motor speeds over the radio. Given that we can manipulate them through the parameter framework, I was wondering how frequently (in terms of Hz) we can change them the motor power commands through this framework. Maybe you can speak to the priority assigned to param read/write inside your firmware.

This info would be useful to see how quickly my controller (that I implement on my laptop to compute the required motor speeds) can react to a deviation from the desired set-point.

Is it worth defining a new packet in the firmware/cflib for this purpose?

PS: I want to point out that according to your documentations [here](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/userguides/logparam/#parameters), it is advised to write parameter values for the sake of initialization only. 

Thanks

**Reply by krichardsson** [reply]
Parameter updates are sent from the computer as soon as possible and the latency/rate depends mainly on how many other packets that are sent. In the crazyflie the new motor value will be updated at 1000Hz. The radio will be the limiting factor here and I tested on my system and got around 200 updates/s. One problem is that there is a [queue in the python lib for outgoing parameter updates](https://github.com/bitcraze/crazyflie-lib-python/blob/ea0154e68ad4ae6a6deda3d4ffb5f1011efed829/cflib/crazyflie/param.py#L604) and if you push updates too often they will stack up and you will get latency. The param system is not really designed for real time updates.

> Is it worth defining a new packet in the firmware/cflib for this purpose?

The parameter updates are acked from the CF (to ensure they are updated) and adding your own packet would bypass this mechanism and reduce latency and increase the rate. It all depends on your needs and your setup. If you are not logging too much you can probably update the motors at around 30-40 Hz with the parameter framework.

> PS: I want to point out that according to your documentations [here](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/userguides/logparam/#parameters), it is advised to write parameter values for the sake of initialization only.

This on the CF side, it means that you should not directly modify a variable in the firmware if it is bound to a paramter.

**Reply by NageshEranki** [reply]
Hi

Thank you so much for your response. I think I have what I needed. I'll stick to the parameter framework for my use case for now. But for future reference, could you elaborate on the following? I don't understand what 'acked' means in this context:

> The parameter updates are acked from the CF (to ensure they are updated) and adding your own packet would bypass this mechanism.

Additionally, if I did want to experiment with designing my own packets, could you shed light (at a birds eye level) on what modifications I would need to make in the firmware/cflib to send my custom packets over the radio?

Thanks

**Reply by krichardsson** [reply]
A good starting point for creating your own packet is the [App channel example](https://github.com/bitcraze/crazyflie-firmware/tree/master/examples/app_appchannel_test)

**Reply by NageshEranki** [reply]
Hi

>  The radio will be the limiting factor here and I tested on my system and got around 200 updates/s. 

I want to test this myself. Are you recording timestamps in your client code each time you receive ACK from the CF?

Thanks

**Reply by krichardsson** [reply]
For log packets there is a time stamp, I'm not sure for generic packets. You can always add it your self when you receive it.

**Reply by knmcguire** [reply]
@NageshEranki If you have any more questions, it's better to start a new discussion thread. This particular thread is pretty old...
